### PR TITLE
Introduce cache objects supporting ES maps

### PR DESCRIPTION
**Pull request by toomuchdesign**
### What kind of change does this PR introduce? *(Bug fix, feature, docs update, ...)*
Feature

### What is the current behaviour? *(You can also link to an open issue here)*
Currently is only possible to store [cache keys of type `string` or `number`](https://github.com/toomuchdesign/re-reselect/blob/v1.0.1/src/index.js#L28). This check is currently made by the **core of the library** itself.

In order to allow the implementation of `cacheObject`s supporting [Maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) or any other cache system, we might consider to find a different way of validating `cacheKey` entries, and let `cacheObject` free to receive any kind of `cacheKey`.

### What is the new behaviour?
This approach consists of letting any `cacheObject` declare a new optional method named `isCacheKeyValid` _(very open to suggestions 😄 )_.

` re-reselect` will call this method to validate the processed `cacheKey` or the default one if no `isCacheKeyValid` method is found.

### Does this PR introduce a breaking change? *(What changes might users need to make in their application due to this PR?)*
Currently no changes, old `cacheObjects` are still relying on default `cacheKey` validation.

This PR might break only if an user used a custom `cacheObject` which declares a `isCacheKeyValid` method.

### Other information:
Not sure whether we should provide a default `cacheKey` validation method.

Still work in progress.

### Please check if the PR fulfills these requirements:
- [x] Tests for the changes have been added
- [ ] Docs have been added / updated

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -21,11 +25,13 @@ export default function createCachedSelector(...funcs) {
...
 
+    const isCacheKeyValid = cache.isCacheKeyValid || defaultCacheKeyValidator;
+
+    // Application receives this function
     const selector = function(...args) {
-      // Application receives this function
       const cacheKey = resolver(...args);
 
-      if (typeof cacheKey === 'string' || typeof cacheKey === 'number') {
+      if (isCacheKeyValid(cacheKey)) {
```
Comment (xsburg) [inline src/index.js]: I would very much like giving a warning if the key is invalid. Generally, it means that something is wrong in my key implementation. In fact, I would consider `undefined` as a suspicious key in all cases (even though Maps support it as a valid key).
```
console.warn(`[re-reselect] Invalid cache key '${cacheKey}' has been returned by resolver function.`);
```
Giving it a second thought, returning `undefined` as the result as a selector might lead to bugs which are very difficult to trace back in this case. Should we maybe throw an exception in this case? I tend to be rather restrictive, just an idea.

File: src/index.d.ts [hunk 2]
```diff
@@ -627,3 +627,20 @@ export class LruCacheObject implements ICacheObject {
   remove (key: string|number): void;
   clear (): void;
 }
+
+export class FlatMapCacheObject implements ICacheObject {
+  set (key: any, selectorFn: any): void;
+  get (key: any): any;
+  remove (key: any): void;
+  clear (): void;
+  isCacheKeyValid (): boolean;
```
Comment (xsburg) [inline src/index.d.ts]: Should take `cacheKey: any` as a parameter. Also, there's the `Resolver` definition in this typings that restricts its return value to `string | number`, should be `any` now.

**Comment by xsburg** [comment]
Hi @toomuchdesign,
Great idea to provide Map-based caches! Dropped a few comments about the code and will fix the typings a bit later.

**Comment by toomuchdesign** [comment]
Hi @xsburg, very happy to hear that it makes sense to you!
I'll keep on working on this and on your suggestions in the next days.

Thank you for reviewing this messy PR,. Feel free to add your ideas as always!
Cheers! 🎉

File: src/cache/FifoMapCacheObject.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  get(key) {
+    return this._cache.get(key);
+  }
+  remove(key) {
+    this._cache.delete(key);
+  }
+  clear() {
+    this._cache.clear();
+  }
+  isCacheKeyValid() {
```
[the remark this answers is not in the document]
Comment (toomuchdesign) [reply src/cache/FifoMapCacheObject.js]: ✅

File: src/cache/FlatMapCacheObject.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    this._cache = new Map();
+  }
+  set(key, selectorFn) {
+    this._cache.set(key, selectorFn);
+  }
+  get(key) {
+    return this._cache.get(key);
+  }
+  remove(key) {
+    delete this._cache.delete(key);
```
[the remark this answers is not in the document]
Comment (toomuchdesign) [reply src/cache/FlatMapCacheObject.js]: ✅

File: src/index.js (hunk 1 above)
Comment (toomuchdesign) [reply src/index.js]: I'd start with a `console.warn`, good point.
Wouldn't be an exception a bit too punitive since we are just deriving data?

Comment (xsburg) [reply src/index.js]: My reasoning for throwing an exception is this: according to current behaviour a cached selector will return `undefined` if the resolved key is invalid. This value will then be used in the user code and likely to throw a `TypeError` exception when accessed:
```javascript
const user = getUserDetailsById(state, userId); // cached with re-reselect by userId
return {
    email: user.email // TypeError thrown
};
```
So the earlier we indicate to the source of the problem, the easier it will be to avoid long debugging trying to find the source of the `undefined`.
To sum up, I would say that warning would be enough if we still returned a correct result and didn't break the user code apart from caching.

File: src/index.js [hunk 5]
```diff
@@ -1,9 +1,13 @@
 import {createSelector} from 'reselect';
 import FlatCacheObject from './cache/FlatCacheObject';
 
-export default function createCachedSelector(...funcs) {
-  const defaultCacheCreator = FlatCacheObject;
+const defaultCacheCreator = FlatCacheObject;
+const defaultCacheKeyValidator = cacheKey => {
```
[the remark this answers is not in the document]
Comment (toomuchdesign) [reply src/index.js]: ✅

File: src/cache/FifoMapCacheObject.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import validateCacheSize from './util/validateCacheSize';
+
+export default class FifoMapCacheObject {
```
Comment (ianobermiller) [inline src/cache/FifoMapCacheObject.js]: Should it be called `FifoCacheMap`? No `Object`?

File: src/cache/FlatMapCacheObject.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+export default class FlatMapCacheObject {
```
Comment (ianobermiller) [inline src/cache/FlatMapCacheObject.js]: Ditto, `FlatCacheMap`?

File: src/cache/LruCacheObject.js [hunk 8]
```diff
@@ -43,4 +38,7 @@ export default class LruCacheObject {
       this._cacheOrdering.splice(index, 1);
     }
   }
+  isValidCacheKey(cacheKey) {
+    return typeof cacheKey === 'string' || typeof cacheKey === 'number';
```
Comment (ianobermiller) [inline src/cache/LruCacheObject.js]: Consider breaking this out into a `isValidObjectKey` function.

Comment (toomuchdesign) [reply src/cache/LruCacheObject.js]: ✅

**Comment by toomuchdesign** [comment]
Hi @xsburg, Hi @ianobermiller,
first of all I want to thank you for contributing on this new chapter of `re-reselect` saga! It's very appreciated 🥇.

I'd like to ask for some advice about a few points.

## Cache names

### Current names
Since now we have **2 cache implementations** for **3 cache types**,  we need a better naming strategy. Cache objects are **currently exposed with the following names**:

- `FlatCacheObject`
- `FifoCacheObject`
- `LruCacheObject`

### Proposed names
Since now we have **2 different cache implementations** using respectively JS **Objects** and **Maps**, I was thinking of:

- `FlatObjectCache`
- `FifoObjectCache`
- `LruObjectCache`
- `FlatMapCachet`
- `FifoMapCache`
- `LruMapCache`

Any comment, different ideas?

### Deprecating old cache exports
In order to keep backward compatibility I'd like to also **keep the old cache exports**.

Do you think it's worth trying to **fire a `console.warn` when cache objects are imported trough the old exports**?

## Deprecated options
6 months ago we deprecated calling `re-reselect` by providing the [`option` parameter as a function](https://github.com/toomuchdesign/re-reselect/blob/v1.0.1/src/index.js#L15).

Since this PR will be delivered with a new major release, we might **consider removing the old behavior** and the deprecation message. Do you think the time is ripe for this, or **6 months is still to early**?

## Anything else?
Looking forward to hearing your feedback!

**Comment by xsburg** [comment]
Hi @toomuchdesign,
I like the ideas. The naming of the cache objects you propose is much clearer and it might be a good idea to introduce it through `console.warn` deprecation warnings.
Regarding `options` deprecation, it might well be the time. 6 months is a lot, the only thing is that I would prefer throwing an error if `options` is a function now. This way we will guarantee that mistakes are not introduced when somebody decides to upgrade and misses the warning.

**Comment by toomuchdesign** [comment]
Hi all, I'd say I've done here! 🎉 
It took more than expected since I've been able to work on it only during random spare moments over  the last weeks. But here we are!

So, the feature is there, covered by test and docs are updated. I'd like to publish next week.

**The only thing left out** is **deprecation warning for the 3 old cache imports**: there is no way of deprecating JS exports and in our case the only solution I found consisted of declaring 3 deprecated cache classes extending the good one just to `console.warn` on construction time.

I decided to not go for this solution, but I'm open to different ideas.

If you had little time to have a look to the PR and drop any suggestion, I'll be happy to receive any feedback.

Greetings!

**Comment by xsburg** [comment]
Hi @toomuchdesign,
Sorry I haven't replied for so long.

Regarding deprecation warnings for the old cache objects, I don't think they are really necessary as long as some of us still use polyfills. Even though the lookup time in polyfills is still normally O(1), there are cases when it's not (for frozen objects used as keys, [source](https://github.com/zloirock/core-js/blob/v2.0.3/modules/_collection-strong.js#L21)). So I'm quite happy with having the plain old FlatObjectCache with all its restrictions by default, no surprises.

All in all looks great, I think we are ready!

Thanks!

**Comment by toomuchdesign** [comment]
Thanks to you, @xsburg. Shall the merge happen!
